### PR TITLE
Fix the issue that max_unpool2d and max_unpool3d cpp signature should…

### PR DESCRIPTION
**Pull request by chengjinfang1216**
… be similar(#39757)

### CODE REVIEW

**Comment by chengjinfang1216** [comment]
@ailzhang,  
could you please review current source code for me?
Note: about moving calculate output_size logic to cpp side, there seems to be a lot of code changes, so it is not included in this PR.

**Comment by chengjinfang1216** [comment]
hello @mruberry, is there any comment about this? I'm very sorry  I have been busy recently and forgot to check this issue, waiting for your reply

**Comment by mruberry** [comment]
> hello @mruberry, is there any comment about this? I'm very sorry I have been busy recently and forgot to check this issue, waiting for your reply

Sorry @chengjinfang1216, looks like this slipped through the cracks. We'll take a look soon!

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 1]
```diff
@@ -118,17 +175,113 @@ Tensor& max_unpooling2d_forward_out_cpu(
...
+    IntArrayRef output_size,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  auto oheight = output_size[0];
+  auto owidth = 1;
+  TORCH_CHECK(output.is_contiguous(), "output must be contiguous");
+  max_unpooling1d_shape_check(
+      self_, Tensor(), indices_, output_size, stride, padding);
+  /* extend one dimension for tensor so that it can use max_unpooling2d_backward_out_cpu_frame method. */
+  auto self = self_.contiguous().unsqueeze(3);
```
Comment (heitorschueroff) [inline aten/src/ATen/native/MaxUnpooling.cpp:253]: The input could be 2D or 3D (batch), so you need to condition the unsqueeze on that. The original code actually has the same issue and does not handle non-batch input correctly. See here https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L694

Could you change both this and the code in functional.py to do the unsqueeze conditionally on whether the input is 2D or 3D?

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 2]
```diff
@@ -118,17 +175,113 @@ Tensor& max_unpooling2d_forward_out_cpu(
...
+Tensor max_unpooling1d_forward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntArrayRef output_size,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  auto output = at::empty({0}, self.options());
+  max_unpooling1d_forward_out_cpu(output, self, indices, output_size, stride, padding);
+  /* reduce one dimension */
+  return output.squeeze(3);
```
Comment (heitorschueroff) [inline aten/src/ATen/native/MaxUnpooling.cpp:282]: Squeeze must be conditioned on input dimension 2D or 3D

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 3]
```diff
@@ -460,13 +599,91 @@ Tensor max_unpooling2d_backward_cpu(
...
+    IntArrayRef stride,
+    IntArrayRef padding) {
   auto grad_input = at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   max_unpooling2d_backward_out_cpu(
-      grad_input, grad_output, self, indices, output_size);
+      grad_input, grad_output, self, indices, output_size, stride, padding);
+  return grad_input;
+}
+
+Tensor& max_unpooling1d_backward_out_cpu(
```
Comment (heitorschueroff) [inline aten/src/ATen/native/MaxUnpooling.cpp:611]: This has the same issues with the forward. Namely, check shape before accessing elements and handle 2D vs 3D input size

File: aten/src/ATen/native/cuda/MaxUnpooling.cu [hunk 4]
```diff
@@ -98,35 +98,219 @@ __global__ void max_unpooling3d_backward_kernel(
   }
 }
 
+static void max_unpooling2d_shape_check(
```
Comment (heitorschueroff) [inline aten/src/ATen/native/cuda/MaxUnpooling.cu:101]: Maybe move this to Pool.h. There's already some utility shape checking code there for other pooling functions. Then call it on both CPU and CUDA

File: aten/src/ATen/native/cuda/MaxUnpooling.cu [hunk 5]
```diff
@@ -98,35 +98,219 @@ __global__ void max_unpooling3d_backward_kernel(
...
+    const Tensor& indices,
+    IntArrayRef output_size,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  auto output = at::empty({0}, self.options());
+  max_unpooling2d_forward_out_cuda(output, self, indices, output_size, stride, padding);
+  return output;
+}
 
+static void max_unpooling1d_shape_check(
```
Comment (heitorschueroff) [inline aten/src/ATen/native/cuda/MaxUnpooling.cu:247]: Same here, move it to Pool.h.

**Review by heitorschueroff — CHANGES_REQUESTED** [review]
This looks great!

I left a lot of comments but they are mostly around 2 issues. One is to do the shape check before accessing the elements in output_size and the other is to handle 2D and 3D cases properly in the new unpool1d code.

You can look at the max_unpool2d for an example on how to handle the 2D vs 3D (there is 4D vs 5D) case properly. I also linked to some original code that does not do this as well, it would be great if you could include a fix for it in this PR.

Looking forward for the updates and let me know if you have any questions.

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 6]
```diff
@@ -76,28 +76,85 @@ Tensor max_unpooling2d_forward_out_cpu_frame(
...
       "Shape of indices should match shape of input");
 
-  TORCH_CHECK(self_.numel() > 0, "Input must be non-empty");
+  TORCH_CHECK(input.numel() > 0, "Input must be non-empty");
+  TORCH_CHECK(
+      stride[0] > 0 && stride[1] > 0,
+      "strides should be greater than zero, but got stride: ",
+      stride);
+
+  int dimw = 2;
```
Comment (mruberry) [inline aten/src/ATen/native/MaxUnpooling.cpp:114]: Do this with an if/else block instead of adding

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 7]
```diff
@@ -460,13 +599,91 @@ Tensor max_unpooling2d_backward_cpu(
...
+
+  /* extend one dimension for tensor so that it can use max_unpooling2d_backward_out_cpu_frame method. */
+  auto indices = indices_.contiguous().unsqueeze(3);
+  auto self_ = self.contiguous().unsqueeze(3);
+
+  /* resize */
+  grad_input.resize_as_(self_);
+  grad_input.zero_();
+
+  if (self_.ndimension() == 4) {
```
Comment (mruberry) [inline aten/src/ATen/native/MaxUnpooling.cpp:642]: Same issue with these additions as above. use an if/else for the number of dimensions to set these properly.

File: aten/src/ATen/native/MaxUnpooling.cpp [hunk 8]
```diff
@@ -460,13 +599,91 @@ Tensor max_unpooling2d_backward_cpu(
...
+    IntArrayRef output_size,
+    IntArrayRef stride,
+    IntArrayRef padding) {
+  TORCH_CHECK(grad_input.is_contiguous(), "grad_input must be contiguous");
+  int64_t oheight = output_size[0];
+  int64_t owidth = 1;
+  int dimw = 2;
+  int dimh = 1;
+  int nbatch = 1;
+  int nslices;
```
Comment (mruberry) [inline aten/src/ATen/native/MaxUnpooling.cpp:625]: Declare these when they're defined, not up front in the function.

File: torch/nn/functional.py [hunk 9]
```diff
@@ -677,12 +677,9 @@ def max_unpool1d(input, indices, kernel_size, stride=None, padding=0,
...
     output_size = _unpool_output_size(input, kernel_size, _stride, padding,
                                       output_size)
-    if isinstance(output_size, list):
-        output_size = output_size + [1]
-    else:
-        output_size = output_size + (1,)
-    return torch._C._nn.max_unpool2d(input.unsqueeze(3), indices.unsqueeze(3),
-                                     output_size).squeeze(3)
+
+    return torch._C._nn.max_unpool1d(input, indices,
```
Comment (mruberry) [inline torch/nn/functional.py:681]: Is this PR combining multiple changes? That is, is this change required to fix the issue?

**Comment by mruberry** [comment]
Hey @chengjinfang1216, thanks for the PR! I have a few questions/concerns:

- The new behavior needs to be tested. The tests can go in test_nn.py/TestNNDeviceType. Let us know if you have questions about writing tests.
- If this change is made, will it break existing C++ programs and serialized Torchscript that was using the existing signatures? This can be tested by scripting a simple function that uses these operations (from Python), saving that script, building with your change, and then verifying the script still runs as expected.
- Is this change as simple as possible? Do the 1d variants of this function need to be added to address the issue?

**Comment by chengjinfang1216** [comment]
> This looks great!
> 
> I left a lot of comments but they are mostly around 2 issues. One is to do the shape check before accessing the elements in output_size and the other is to handle 2D and 3D cases properly in the new unpool1d code.
> 
> You can look at the max_unpool2d for an example on how to handle the 2D vs 3D (there is 4D vs 5D) case properly. I also linked to some original code that does not do this as well, it would be great if you could include a fix for it in this PR.
> 
> Looking forward for the updates and let me know if you have any questions.

Hi, @heitorschueroff ， appreciate for your kind reply， I’ll take a look for detail.

**Comment by heitorschueroff** [comment]
@chengjinfang1216 Are you still interested in pursuing this PR? If so, let me know so I can take another look and help you if you  are blocked in anyway. We really appreciate this contribution to the project.

**Comment by shubhambhokare1** [comment]
Was wondering, if there is going to be any work on this.
I am currently working on adding maxUnpool support to ONNX exporter as in the PR #55411

File: aten/src/ATen/native/native_functions.yaml [hunk 10]
```diff
@@ -6409,26 +6409,52 @@
...
 
-- func: max_unpool2d_backward.grad_input(Tensor grad_output, Tensor self, Tensor indices, int[2] output_size, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: max_unpool2d_backward.grad_input(Tensor grad_output, Tensor self, Tensor indices, int[2] output_size, int[2] stride, int[2] padding, *, Tensor(a!) grad_input) -> Tensor(a!)
   python_module: nn
   dispatch:
     CPU: max_unpooling2d_backward_out_cpu
     CUDA: max_unpooling2d_backward_out_cuda
 
-- func: max_unpool2d_backward(Tensor grad_output, Tensor self, Tensor indices, int[2] output_size) -> Tensor
+- func: max_unpool2d_backward(Tensor grad_output, Tensor self, Tensor indices, int[2] output_size, int[2] stride, int[2] padding) -> Tensor
```
Comment (shubhambhokare1) [inline aten/src/ATen/native/native_functions.yaml:6457]: Would it be possible to add kernel_size as a param to these signatures?

**Pull request closed without merging**